proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,321 @@
+* Integration with 3rd party asynchronous APIs. This is necessary to handle executions that are completed at some time in the future, where retries will need to be explicitly scheduled from within the asynchronous execution. This is common when working with various 3rd party asynchronous tools such as Netty, RxJava, Vert.x, etc.
+* Require immutable failure handling policy configuration
+* Some Failure policy configurations, e.g. CircuitBreaker, RetryPolicy, can be used stand alone. For example, it has been very useful for circuit breakers to be standalone constructs which can be plugged into and intentionally shared across multiple executions. Likewise for retry policies. Additionally, an Execution construct can be offered that allows retry policies to be applied to some logic in a standalone, manually controlled way.
+
+Advanced requirements:
+
+* Event: Since this approach to fault tolerance involves handing execution over to some foreign code, it's very useful to be able to learn when executions are taking place and under what circumstances (onRetry, onFailedAttempt, onFailure, etc).
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/ezFC1TLGozU)
+
+## Motivation
+
+Currently there are at least two libraries to provide fault tolerance. It is best to uniform the technologies and define a standard so that micro service applications can adopt and the implementation of fault tolerance can be provided by the containers if possible.
+
+## Proposed solution
+
+Separate the responsibility of executing logic (Runnables/Callables/etc) from guiding when execution should take place (through retry policies, bulkheads, circuit breakers). In this way, failure handling strategies become configuration that can influence executions, and the execution API itself is just responsible for receiving some configuration and performing executions.
+
+By default, a failure handling strategy could assume, for example, that any exception is a failure. But in some cases, a `null` or negative return value could also be considered a failure. Users should be able to define this, and a user's definition of a failure is what should influence execution. (This all is what the Failsafe RetryPolicy's `retryOn`, `retryWhen`, `abortIf`, etc methods are all about - defining a failure).
+
+Standardise the Fallback, Bulkhead and CircuitBreaker APIs and provide implementations.
+
+* CDI-first approach to apply RetryPolicy, Fallback, BulkHead, CircuitBreaker using annotations
+
+## Detailed design (One example of implemenations)
+
+
+### CDI-based approach 
+Use interceptor and annotation to specify the execution and policy configration.
+An annotation of Asynchronous has to be specified for any asynchronous calls. Otherwise, synchronous execution is assumed. 
+The implementation should provide two interceptors, one for synchronous invocation and the other for asynchronous invocation. 
```

Comment:
If synchronous and asynchronous execution is desired in one bean, does this imply a need for double injection ? Once with the async interceptor and once with the sync interceptor ?